pkg/location/reference.go

```diff
@@ -0,0 +1,13 @@
+package location
+
+import "fmt"
+
+type Reference interface {
+	fmt.Stringer
```

Comment:
Should each reference be expressible as a URI?

file://
oci://
https://github.com/...

Most certainly seem to be; if not all, maybe we need an interface that can go back and forth from single-string (URI) representation?

Reply:
Ah, that is true, isn't it? Though TBH the oci:// one isn't strictly canonical... It is the same format helm uses for oci, but really it's pulled out of thin air.

Come to think of it - the Stringer isn't also strictly round tripping with parse. Instead it's more like a "%v" property-fmt style dump. Maybe it should be treated more like the inverse of parse?

I suppose a question would be - you'd probably still want to have non-file:// dir for CLI args --- if we did add a URI is there a place you're imagining that where URI-valued-string would be the standard ref notation?

Reply:
Oh! Yes, agreed. In the next PR where some of this is used under `kpt pkg get` there are methods that are aiming to be in the kptfileutil package. It's possible there could be a better home for those, but that's where methods like `kptfileutil.UpdateUpstreamLockFromGit` are today so seemed like a good starting point.

```go
ref, err := kptfileutils.NewReferenceFromUpstream(kf)
refLock, err := kptfileutils.NewReferenceLockFromUpstreamLock(kf)
kf.Upstream, err = kptfileutils.NewUpstreamFromReference(ref)
kf.UpstreamLock, err = kptfileutils.NewUpstreamLockFromReferenceLock(refLock)
```